### ISSUE TITLE
How do I find which entity (or entities) are throwing a circular dependency when reinitialising a scaffold?

### ISSUE THREAD

**Issue by seanjbsol** [dotnet/efcore#27053]
Hi all

I'm looking to create a new migration for a schema that has evolved considerably since I started the project. The main reason behind this is because I'm looking to move from Sqlite to SQL Server to allow other reporting services and a more robust method of running queries against the dataset when the application is live.

The scaffold operation throws an error when I clear all the migrations data and start again. I've tried to correct a number of the entities, but I'm not sure which one (or ones) are specifically causing the issue.

I cannot rewrite the whole schema as I have a database in production with the current schema, so I need to achieve this by finding the edge cases and fixing those. Which I thought I'd done - but still getting the error. Is there any way to get specific details as to which POCOs are causing the circular dependency specifically?

### Include stack traces

Include the full exception message and stack trace for any exception you encounter.

Use triple-tick fences for stack traces. For example:

```
Microsoft.EntityFrameworkCore.Migrations.Operations.CreateTableOperation'.
PM> add-migration Initial -Verbose
Using project 'FleetServApp'.
Using startup project 'FleetServApp'.
Build started...
Build succeeded.
C:\Program Files\dotnet\dotnet.exe exec --depsfile S:\src\fleetserv.app\BlazorEngineeringServices\bin\Debug\net6.0\FleetServApp.deps.json --additionalprobingpath C:\Users\seanj\.nuget\packages --additionalprobingpath "C:\Program Files (x86)\Microsoft Visual Studio\Shared\NuGetPackages" --runtimeconfig S:\src\fleetserv.app\BlazorEngineeringServices\bin\Debug\net6.0\FleetServApp.runtimeconfig.json C:\Users\seanj\.nuget\packages\microsoft.entityframeworkcore.tools\6.0.1\tools\netcoreapp2.0\any\ef.dll migrations add Initial --json --verbose --no-color --prefix-output --assembly S:\src\fleetserv.app\BlazorEngineeringServices\bin\Debug\net6.0\FleetServApp.dll --project S:\src\fleetserv.app\BlazorEngineeringServices\FleetServApp.csproj --startup-assembly S:\src\fleetserv.app\BlazorEngineeringServices\bin\Debug\net6.0\FleetServApp.dll --startup-project S:\src\fleetserv.app\BlazorEngineeringServices\FleetServApp.csproj --project-dir S:\src\fleetserv.app\BlazorEngineeringServices\ --language C# --configuration Debug --working-dir S:\src\fleetserv.app --root-namespace FleetServApp
Using assembly 'FleetServApp'.
Using startup assembly 'FleetServApp'.
Using application base 'S:\src\fleetserv.app\BlazorEngineeringServices\bin\Debug\net6.0'.
Using working directory 'S:\src\fleetserv.app\BlazorEngineeringServices'.
Using root namespace 'FleetServApp'.
Using project directory 'S:\src\fleetserv.app\BlazorEngineeringServices\'.
Remaining arguments: .
Finding DbContext classes...
Finding IDesignTimeDbContextFactory implementations...
Finding application service provider in assembly 'FleetServApp'...
Finding Microsoft.Extensions.Hosting service provider...
Using environment 'Development'.
Using application service provider from Microsoft.Extensions.Hosting.
Found DbContext 'SqliteContext'.
Finding DbContext classes in the project...
Using context 'SqliteContext'.
Microsoft.EntityFrameworkCore.Infrastructure[10403]
      Entity Framework Core 6.0.1 initialized 'SqliteContext' using provider 'Microsoft.EntityFrameworkCore.Sqlite:6.0.1' with options: using lazy loading proxies MigrationsAssembly=FleetServApp 
Finding design-time services referenced by assembly 'FleetServApp'...
Finding design-time services referenced by assembly 'FleetServApp'...
No referenced design-time services were found.
Finding design-time services for provider 'Microsoft.EntityFrameworkCore.Sqlite'...
Using design-time services from provider 'Microsoft.EntityFrameworkCore.Sqlite'.
Finding IDesignTimeServices implementations in assembly 'FleetServApp'...
No design-time services were found.
System.InvalidOperationException: Unable to save changes because a circular dependency was detected in the data to be saved: 'Microsoft.EntityFrameworkCore.Migrations.Operations.CreateTableOperation ->
Microsoft.EntityFrameworkCore.Migrations.Operations.CreateTableOperation ->
Microsoft.EntityFrameworkCore.Migrations.Operations.CreateTableOperation ->
Microsoft.EntityFrameworkCore.Migrations.Operations.CreateTableOperation'.
   at Microsoft.EntityFrameworkCore.Utilities.Multigraph`2.ThrowCycle(List`1 cycle, Func`2 formatCycle, Func`2 formatException)
   at Microsoft.EntityFrameworkCore.Utilities.Multigraph`2.TopologicalSort(Func`4 tryBreakEdge, Func`2 formatCycle, Func`2 formatException)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationsModelDiffer.Sort(IEnumerable`1 operations, DiffContext diffContext)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationsModelDiffer.GetDifferences(IRelationalModel source, IRelationalModel target)
   at Microsoft.EntityFrameworkCore.Migrations.Design.MigrationsScaffolder.ScaffoldMigration(String migrationName, String rootNamespace, String subNamespace, String language)
   at Microsoft.EntityFrameworkCore.Design.Internal.MigrationsOperations.AddMigration(String name, String outputDir, String contextType, String namespace)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.AddMigrationImpl(String name, String outputDir, String contextType, String namespace)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.AddMigration.<>c__DisplayClass0_0.<.ctor>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.<>c__DisplayClass3_0`1.<Execute>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.Execute(Action action)
Unable to save changes because a circular dependency was detected in the data to be saved: 'Microsoft.EntityFrameworkCore.Migrations.Operations.CreateTableOperation ->
Microsoft.EntityFrameworkCore.Migrations.Operations.CreateTableOperation ->
Microsoft.EntityFrameworkCore.Migrations.Operations.CreateTableOperation ->
Microsoft.EntityFrameworkCore.Migrations.Operations.CreateTableOperation'.
```
### Include provider and version information

EF Core version:6.0.1
Database provider: Microsoft.EntityFrameworkCore.SqlServer & Microsoft.EntityFrameworkCore.Sqlite
Target framework: .NET 6.0
Operating system: Windows
IDE: Visual Studio 2022

**Comment by VicenzoMartinelli** [comment]
I'm facing the same issue in a solution with 700+ entities. 
At the time it's blocking my migration from .net core 3.1 to .net6.

**Comment by VicenzoMartinelli** [comment]
Any news about this issue?

**Comment by VicenzoMartinelli** [comment]
some related questions on stackoverflow:
https://stackoverflow.com/questions/70018530/ef-core-6-system-invalidoperationexception-unable-to-save-changes-because-a-c
https://stackoverflow.com/questions/70012349/self-referencing-entity-circular-dependency-happening-just-on-ef-core-6-0

**Comment by ajcvickers** [comment]
The underlying issue here is tracked by #26834. This has been fixed and merged for the 6.0.2 release. You should be able to get the fix now by using a [daily build](https://github.com/dotnet/efcore/blob/main/docs/DailyBuilds.md). If the daily build doesn't fix the issue, then please attach a small, runnable project or post a small, runnable code listing that reproduces what you are seeing so that we can investigate.

@AndriySvyryd Is there a way to identify the entities involved?

**Comment by AndriySvyryd** [comment]
This exception was never meant to be thrown in this scenario, that's why the message doesn't have the specific table names that `CreateTableOperation` were created for. Look for tables that have foreign key cycles, usually it's caused by a foreign key from table A to B and a foreign key from table B to A. Once you remove those tables or foreign keys and don't get the exception you should be able to add them back in a separate migration.
